### ISSUE TITLE
Warning: preg_match(): Compilation failed: unrecognized character

### ISSUE THREAD

**Issue by jrfnl** [maglnet/ComposerRequireChecker#416]
For select repos, when I run the tool from a PHAR file on Windows 10, I receive the following PHP warning (x 40 or so):|
```
Warning: preg_match(): Compilation failed: unrecognized character follows \ at offset 19 in phar://path/to/composer-require-checker.4.7.0.phar/src/ComposerRequireChecker/FileLocator/LocateAllFilesByExtension.php on line 56
```

Runtime details:
```
> composer-require-check check composer.json
PHP 8.2.8 (cli) (built: Jul  4 2023 15:53:15) (ZTS Visual C++ 2019 x64)
Copyright (c) The PHP Group
Zend Engine v4.2.8, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.8, Copyright (c), by Zend Technologies

ComposerRequireChecker 4.7.0@a80a012ba94d5423b02754b2816960a74d3d4f3f

Warning: preg_match(): Compilation failed: unrecognized character follows \ at offset 19 in phar://path/to/composer-require-checker.4.7.0.phar/src/ComposerRequireChecker/FileLocator/LocateAllFilesByExtension.php on line 56
```

I suspect the issue is with the regex escaping of paths in a blacklist, but not sure how to debug this with just a PHAR file.

Neither the path to the repo, nor the path to the PHAR file contains spaces in any of the directory paths

https://github.com/maglnet/ComposerRequireChecker/blob/4.8.x/src/ComposerRequireChecker/FileLocator/LocateAllFilesByExtension.php#L73

**Comment by maglnet** [comment]
Hi, maybe you could provide the config files or the regexes, that have been used when this error occurred?

**Comment by jrfnl** [comment]
@maglnet Wish I could, but I wasn't using any.

I just ran the following quite plain command:
```bash
composer-require-check check composer.json
```

**Comment by maglnet** [comment]
OK, could you provide the composer.json file, maybe there's something within the autoload sections that may help debugging?

**Comment by heiglandreas** [comment]
From what I read in that function multiple immediately following occurrences of a directory separator are replaced by a single directory separator.

But that one is not escaped when it is a windows backslash.

It looked like there is a preg_quote missing somewhere...

Which would fit with the error message

But I might also have missed it...

**Comment by jrfnl** [comment]
@heiglandreas Yes, I was thinking along the same lines, but as I wasn't providing a blacklist, I was unsure how to figure out which character was not being escaped properly.

Either way, just checked and the "problem" repo is public. Here is the `composer.json`: https://github.com/Yoast/wp-test-utils/blob/develop/composer.json

**Comment by jrfnl** [comment]
(Oh and just now ran into the same issue with another repo, though that one is `private` unfortunately. DIfferent offset, same line and even more warnings, but still showing output.)

**Comment by maglnet** [comment]
@jrfnl Since I do not have a Windows system available for quick debugging, would you be able to give the absolute path, where you saved the repo and maybe try to run it within a very "secure" path like `c:\Temp\repo\` or similar, to check if there's something outside the repository path that causes this error?

**Comment by jrfnl** [comment]
The local path to the repo is: `D:\000_GitHub\WordPress_SEO_by_Yoast\wp-test-utils`.

I have been scanning other repos in the `D:\000_GitHub\WordPress_SEO_by_Yoast` directory as well. Most without issues, a few seeing the same issues as reported here, so I don't think the base path _as such_ is a problem.

**Comment by maglnet** [comment]
Thank you for your input, the base path really does not look suspicious 😀
I'll see if I can reproduce this somehow.

**Comment by jrfnl** [comment]
Well, the only suspicious thing in the base path is the `\0` which could be seen as a null byte in certain circumstances 😂 , but as that isn't giving any problems with the other repos in that path, I suspect that isn't the cause of the issue.

I'm a bit busy now, but if needs be, I can clone this repo and do some debug var_dumps if it would help.

**Comment by fredden** [comment]
I've downloaded an evaluation virtual machine from Microsoft and reproduced the bug there. From what I can tell, the problem is here https://github.com/maglnet/ComposerRequireChecker/blob/451403683a2a6509500b619e9eb59037579cdc7f/src/ComposerRequireChecker/FileLocator/LocateAllFilesByExtension.php#L82-L89
where we replace the safe-for-regular-expressions `$dirSep` with a plain `DIRECTORY_SEPARATOR`.

I've unwrapped this complex one-liner into its separate statements, and when I comment out this line:
`$path = preg_replace('{' . $dirSep . '+}', DIRECTORY_SEPARATOR, $path);`
it all seems to work as expected on Windows and Linux.

I'm working through the testsuite now and will open a pull request shortly.

**Comment by Ocramius** [comment]
Handled by @fredden in #419

**Comment by jrfnl** [comment]
Thank you @fredden and @Ocramius for getting this fixed & released.

I've just tested the new PHAR and can confirm that the warnings no longer show.

The output of the tool, however is also different.....
![image](https://github.com/maglnet/ComposerRequireChecker/assets/663378/53724e18-db95-4b0f-b472-048c0530d7cd)

As I haven't used this tool that often yet, I find it hard to tell which of the two outputs would be the "expected" (correct) output, but the difference in the lists is significant.

Should I open a separate issue about this ? Or is the output generated with the 4.7.1 PHAR the expected output ?

**Comment by jrfnl** [comment]
Note: I've just tested with some other projects which didn't show the warnings and for those I get the same output with both 4.7.0 and 4.7.1, so I presume the above is a non-issue. Just wanted to check to make sure.

**Comment by Ocramius** [comment]
Please do indeed make a new issue, if you believe it is indeed a problem
(haven't checked myself, just answering notifications)
